add script project

josemanuel167 opened the pull request:
Add nodeJs script to generate picklist options from RxNorm .txt

josemanuel167 committed:
add script project

josemanuel167 committed:
replace only single quotes

ssettipalli commented:
Looks like script generated output option stableIDs are numerics like below
{
      "stableId": 1000000,
      "optionLabelTemplate": {
        "templateType": "TEXT",
        "templateText": "TRIBENZOR Oral Pill"
      }
    },
    {
      "stableId": 1000001,
      "optionLabelTemplate": {
        "templateType": "TEXT",
        "templateText": "amLODIPineHydrochlorothiazideOlmesartan Oral Pill"
      }
    },
    
I believe we need better stableIDs that reflect option text.
May be user one word or combo or two words capitalized TRIBENZOR_ORAL OR TRIBENZOR_ORAL_PILL

bskinner commented:
The `dist` directory seems to contain compiled code- does this need to be committed to the repository?

ssettipalli commented:
Script should take input / output file names as parameters and remove pushing actual input and output data files to repo.

josemanuel167 committed:
remove /dist from repo

josemanuel167 commented:
@ssettipalli I updated the script to take input/output and removed them from repo. For stable IDs I'm not sure if we should create our owns, adding @MocanaAtBroad for help with this.
Removed /dist from repo cc..@bskinner

bskinner commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,41 @@
+import { RxNormDrugOption } from "./models";
+import { parseToHOCON, writeFile } from "./utils/common";
+
+const inputFileName = 'RxTerms202202.txt';
+const outputFileName = 'output.txt'
```
Let's take the files from the command line instead of hard-coding the paths.

jm167 replied:
Sounds good. Added

jm167 replied:
Readme added

bskinner reviewed (commented):
Looks good- adding two items to address from my end:
* Let's add a short README describing, at a high level, what the script does and how to invoke it.
* The input and output filenames should not be hard coded into the script.

josemanuel167 committed:
choose file from console and recive output name

josemanuel167 committed:
add readme

bskinner commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,72 @@
+
+inquirer.registerPrompt("file-tree-selection", inquirerFileTreeSelection);
+
+inquirer
+  .prompt([
+    {
+      root: "./",
+      type: "file-tree-selection",
+      name: "files",
+      message: "choose RxNorm .txt file"
+    },
+    {
+      type: "input",
+      name: "outputName",
+      message: "Enter output file name. File will be generated on TXT extension"
+    }
+  ])
+  .then((answers: InputOutput) => {
```
This is going to be difficult to incorporate into any sort of non-interactive environment- We should not be prompting for input. If it helps, `npm run` allows for command line arguments to be passed using the `npm run <command> [<args>]` format, where `<args>` is one or more key-value pairs. Any kv pairs provided in `args` will be exposed inside `process.env` under a property named `npm_config_$arg`. For example, `--foo=bar` will be exposed via `process.env.npm_config_foo` and be set to the value `bar`.

If this doesn't work, positional arguments via `process.argv` would be acceptable as well (`npm run start -- <inputPath> <outputPath>`).

Let me know if you have any questions.

josemanuel167 replied:
Added arguments from console and defaults

josemanuel167 committed:
take args from console

josemanuel167 committed:
update readme

josemanuel167 committed:
remove dependencies

josemanuel167 committed:
Merge branch 'develop' into feature/add-script-to-generate-picklist-options

bskinner commented:
Thank you for the changes- let's get this in 🙂

bskinner reviewed (dismissed):
Looks good- let's get this in 🙂

josemanuel167 committed:
update output to include tranlastions from console

josemanuel167 committed:
update read me

josemanuel167 committed:
remove inquirer

MocanaAtBroad commented on scripts/ddp-RxTerms-script/tsconfig.json:
```diff
@@ -0,0 +1,101 @@
+    // "skipDefaultLibCheck": true,                      /* Skip type checking .d.ts files that are included with TypeScript. */
+    "skipLibCheck": true                                 /* Skip type checking all .d.ts files. */
+  }
+}
```
Had a difficult time with this configuration. Please consider replacing with this more concise one that worked for me:
```json
{
  "compilerOptions": {
    "lib": ["es2021"],
    "module": "commonjs",
    "target": "es2021",
    "types": [
      "node"
    ],
    "outDir": "dist",
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true
  }
}

josemanuel167 replied:
This was a fresh file autogenerated, it was nice to know all available options but sure I can replace it

MocanaAtBroad replied:
The options above I mostly got from the nodejs development team. Just types is added.

MocanaAtBroad commented on scripts/ddp-RxTerms-script/package.json:
```diff
@@ -0,0 +1,19 @@
+    "@types/node": "^17.0.21"
+  },
+  "dependencies": {
+    "hocon-parser": "^1.0.1",
```
You don't need hocon-parser. Have some suggetions in code.

MocanaAtBroad commented on scripts/ddp-RxTerms-script/package.json:
```diff
@@ -0,0 +1,19 @@
+  "license": "ISC",
+  "devDependencies": {
+    "@types/node": "^17.0.21"
+  },
```
Let's add Typescript

```suggestion
      "typescript": "^4.4.4"
  },
```

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,53 @@
+}},`;
+  return (prev += stringBuilder);
+}, "");
+
```
All the stuff above I don't think makes sense to do.
You are creating here a gigantic HOCON string, but then, when you write it out, you convert to an object (using HOCON parser) and THEN to JSON string. 
So complicated AND we throw the work away.

josemanuel167 replied:
Not really. Hocon parser library allows to provide strings without commas or quotes: 
![image](https://user-images.githubusercontent.com/98110251/162110820-dce09a11-8877-45a8-a360-085c2efaf5e9.png)
I parse it to HOCON just to double check, and add all needed symbols.
Not creating a HOCON's string particularly just a string,

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,53 @@
+    templateText: option.displayName
+  }
+}));
+
```
I strongly suggest go directly from your rxNormArray to the list of picklistOptions that has the exact shape needed in the configuration and that is what you use to generate the JSON to export, e.g.:
```typescript
const pickListOptions = rxNormArray.map((drug) => ({
    stableId: drug.RXCUI.toString(),    // <--- I noticed we were exporting numbers instead of strings!
    optionLabelTemplate: {
        templateType: 'TEXT',
        templateText: '$drug_name',
        variables: [{
            name: 'drug_name',
            translations: languageCodes.map(lang => ({ language: lang, text: drug.DISPLAY_NAME.replace(/'/g, '') }))
        }]
    }
}));

MocanaAtBroad replied:
You can then do something like:

JSON.stringify({pickListOptions: pickListOptions}, null, 2) 

and that is the contents of the file. One iteration over data, no HOCON generation or parseHOCON necessary.

josemanuel167 replied:
Sounds good, thanks for that. I'm not familiar with ocon and wanted make sure it had the right format.

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,53 @@
+const inputFileName = process.argv[2] || "RxTerms202202.txt";
+const outputFileName = process.argv[3] || "output.txt";
+
+const translations: string[] =  process.argv.filter(arg => ['en', 'es', 'ru'].includes(arg));
```
I think we are unnecessarily limiting the the codes that are acceptable.
Just do something like this:
```typescript
const languageCodes: string[] | void = process.argv[4]?.split(",");

if (!languageCodes) {
    console.error("Comma separated language separators were not provided");
    process.exit(1)
}

Now users can say en,es,fn,etc and we will use it.

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,53 @@
+import { RxNormDrugOption } from "./models";
+import { parseToHOCON, writeFile } from "./utils/common";
+
+const inputFileName = process.argv[2] || "RxTerms202202.txt";
```
I recommend checking that the data file actually exists. So something like this:
```typescript

import * as fs from 'fs';

if (!fs.existsSync(inputFileName)) {
    console.error(`Input file with name ${inputFileName} does not exist`);
    process.exit(1);
}

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/utils/common.ts:
```diff
@@ -0,0 +1,12 @@
+export const parseToHOCON = require("hocon-parser");
```
Not needed. You can delete.

josemanuel167 replied:
Please read explanation above

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/utils/common.ts:
```diff
@@ -0,0 +1,12 @@
+    }
+    console.log("The file was saved!");
+  });
+};
```
Consider just getting rid of this file and move write file to index.ts . We are just calling fs.writefile. Up to you.

josemanuel167 replied:
Agree, removed

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/utils/common.ts:
```diff
@@ -0,0 +1,12 @@
+export const parseToHOCON = require("hocon-parser");
+
+const fs = require("fs");
```
You can do this TypeScript style:
```typescript
import * as fs from 'fs';

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,53 @@
+
+const workSheet = workBook.Sheets[workBook.SheetNames[0]];
+
+const rxNormArray = XLSX.utils.sheet_to_json(workSheet) as RxNormDrugOption[];
```
The parsing seems to work well! The one problem I am seeing is that it converts the drug ids to numbers because they all look like integers. But we need strings for stable ids. Suggested way to handle it below in the suggest code.

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,53 @@
+
+writeFile(
+  outputFileName,
+  JSON.stringify(parseToHOCON(`picklistOptions : [${picklistString}]`), null, 2)
```
I suggest above we just build the needed Javascript object and we stringify/ no HOCON.

josemanuel167 committed:
update requiere to import

josemanuel167 committed:
Update scripts/ddp-RxTerms-script/package.json

Co-authored-by: MocanaAtBroad <mocana@broadinstitute.org>

josemanuel167 committed:
update suggested changes

josemanuel167 committed:
remove common

josemanuel167 committed:
update tsconfig

ssettipalli commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,33 @@
+
+const rxNormArray = XLSX.utils.sheet_to_json(workSheet) as RxNormDrugOption[];
+
+const pickListOptions = rxNormArray.map((drug) => ({
```
In the generated output it should be picklistOptions (small `l`)

ssettipalli commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -0,0 +1,33 @@
+  }
+}));
+
+fs.writeFile(outputFileName, JSON.stringify({pickListOptions}, null, 2), err => {
```
In the generated output it should be picklistOptions (small `l`)

josemanuel167 committed:
rename picklist options

josemanuel167 committed:
filter unique dispay names and generate stableIDs

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -1,33 +1,89 @@
+    .replace(/[^a-zA-Z0-9 ]/g, "")
+    .replaceAll(" ", "_")
+    .toUpperCase();
+
```
I see that a lot of entries use the "\", e.g.:
```
Capsaicin/Lidocaine/Menthol/Methyl salicylate
```
Current implementation just gets rid of the "/". I think would be easier to read if we replace the / for a _ in the final string.
So would like to see something like:
```
CAPSAICIN_LIDOCAINE_MENTHOL_METHYL_SALICYLATE
```

MocanaAtBroad reviewed (commented):
Looks good. I am requesting a small enhancement.

MocanaAtBroad commented on scripts/ddp-RxTerms-script/src/index.ts:
```diff
@@ -1,33 +1,89 @@
+);
+
+const picklistOptions = filteredRxNormsArray.map((drug) => ({
+  stableId: generateStableID(drug.DISPLAY_NAME),
```
I think you are going to have to check that the generated stable id is really unique/cannot count on the current transformation giving you a unique ids.

josemanuel167 committed:
append suffix to truncated stableIDs

josemanuel167 committed:
fix for max lenght before trim

josemanuel167 committed:
remove trim after append index

josemanuel167 committed:
remove 3 chars for 2 digits suffix

ssettipalli reviewed (approved):
tested it & looks good to me.

josemanuel167 committed:
Merge branch 'develop' into feature/add-script-to-generate-picklist-options
